server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocatorTests.java

```diff
@@ -651,6 +654,105 @@ public void testResetDesiredBalanceOnNoLongerMaster() {
+                singleShutdownMetadataBuilder.setGracePeriod(TimeValue.MAX_VALUE);
+            }
+            final var nodeShutdownMetadata = new NodesShutdownMetadata(Map.of(node2.getId(), singleShutdownMetadataBuilder.build()));
+
+            clusterState = ClusterState.builder(clusterState)
+                .metadata(Metadata.builder(clusterState.metadata()).putCustom(NodesShutdownMetadata.TYPE, nodeShutdownMetadata))
+                .build();
+            ClusterServiceUtils.setState(clusterService, clusterState);
+
+            assertTrue(desiredBalanceAllocator.getProcessedNodeShutdowns().isEmpty());
+            rerouteAndWait(service, clusterState, "reroute-after-shutdown");
+            assertTrue("desired balance reset should be called on node shutdown", resetCalled.get());
+            assertThat(desiredBalanceAllocator.getProcessedNodeShutdowns(), equalTo(Set.of(node2.getId())));
+
+            resetCalled.set(false);
+            rerouteAndWait(service, clusterState, "random-reroute");
+            assertFalse("desired balance reset should not be called again for processed shutdowns", resetCalled.get());
+            assertThat(desiredBalanceAllocator.getProcessedNodeShutdowns(), equalTo(Set.of(node2.getId())));
+
+            final var clusterStateBuilder = ClusterState.builder(clusterState)
+                .metadata(Metadata.builder(clusterState.metadata()).putCustom(NodesShutdownMetadata.TYPE, NodesShutdownMetadata.EMPTY));
+            final var nodeRemovedFromCluster = randomBoolean();
+            if (nodeRemovedFromCluster) {
+                clusterStateBuilder.nodes(DiscoveryNodes.builder().add(node1).localNodeId(node1.getId()).masterNodeId(node1.getId()));
+            }
+            clusterState = clusterStateBuilder.build();
+            ClusterServiceUtils.setState(clusterService, clusterState);
+            rerouteAndWait(service, clusterState, "random-reroute");
+            if (nodeRemovedFromCluster) {
+                assertFalse("desired balance reset should not be called again for processed shutdowns", resetCalled.get());
+            } else {
+                assertTrue("desired balance reset should be called again for processed shutdowns", resetCalled.get());
+            }
+            assertTrue(desiredBalanceAllocator.getProcessedNodeShutdowns().isEmpty());
+
+            resetCalled.set(false);
+            if (nodeRemovedFromCluster == false) {
+                clusterState = ClusterState.builder(clusterState)
+                    .nodes(DiscoveryNodes.builder().add(node1).localNodeId(node1.getId()).masterNodeId(node1.getId()))
+                    .build();
+                ClusterServiceUtils.setState(clusterService, clusterState);
+                rerouteAndWait(service, clusterState, "reroute-after-node-removed");
+                assertFalse("desired balance reset should not be called", resetCalled.get());
+                assertTrue(desiredBalanceAllocator.getProcessedNodeShutdowns().isEmpty());
```

Comment:
```suggestion
                assertThat(desiredBalanceAllocator.getProcessedNodeShutdowns(), empty());
```

This way if assertion fails we would get collection contents in the message that could make it easier to troubleshoot the test without adding extra logs if it ever becomes flaky.

Reply:
:+1: see dd521741bf5c0ba0ff64590eea1d6bfe7365a484